fvm/src/call_manager/default.rs

```diff
@@ -568,3 +589,40 @@ where
         res
     }
 }
+
+/// Stores events in layers as they are emitted by actors. As the call stack progresses, when an
+/// actor exits normally, its events should be merged onto the previous layer (merge_last_layer).
+/// If an actor aborts, the last layer should be discarded (discard_last_layer). This will also
+/// throw away any events collected from subcalls (and previously merged, as those subcalls returned
+/// normally).
+#[derive(Default)]
+pub struct EventsAccumulator(Vec<StampedEvent>, Vec<usize>);
+
+impl EventsAccumulator {
+    fn append_event(&mut self, evt: StampedEvent) {
+        self.0.push(evt)
+    }
+
+    fn create_layer(&mut self) {
+        self.1.push(self.0.len());
+    }
+
+    fn merge_last_layer(&mut self) {
+        self.1
+            .pop()
+            .expect("no index in the event accumulator when calling merge_last_layer");
```

Comment:
I'd prefer to return fatal errors here (what we do in the state-tree). Panics tend to poison the entire machine.

Reply:
Converted all to errors except for `finish()`. The usage site expects infallibility. I'll just drop the assert in that case.